flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -538,51 +542,61 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
-
-							Preconditions.checkState(
-								checkpoint != null || throwable != null,
-								"Either the pending checkpoint needs to be created or an error must have been occurred.");
-
-							if (throwable != null) {
-								// the initialization might not be finished yet
-								if (checkpoint == null) {
-									onTriggerFailure(request, throwable);
-								} else {
-									onTriggerFailure(checkpoint, throwable);
-								}
+			FutureUtils.waitForAll(asList(masterStatesComplete, coordinatorCheckpointsComplete))
+				.handleAsync(
+					(ignored, throwable) -> {
+						final PendingCheckpoint checkpoint =
+							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
+
+						Preconditions.checkState(
+							checkpoint != null || throwable != null,
+							"Either the pending checkpoint needs to be created or an error must have been occurred.");
+
+						if (throwable != null) {
+							// the initialization might not be finished yet
+							if (checkpoint == null) {
+								onTriggerFailure(request, throwable);
 							} else {
-								if (checkpoint.isDiscarded()) {
-									onTriggerFailure(
-										checkpoint,
-										new CheckpointException(
-											CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE,
-											checkpoint.getFailureCause()));
-								} else {
-									// no exception, no discarding, everything is OK
-									final long checkpointId = checkpoint.getCheckpointId();
-									snapshotTaskState(
-										timestamp,
-										checkpointId,
-										checkpoint.getCheckpointStorageLocation(),
-										request.props,
-										executions,
-										request.advanceToEndOfTime);
-
-									coordinatorsToCheckpoint.forEach((ctx) -> ctx.afterSourceBarrierInjection(checkpointId));
-
-									onTriggerSuccess();
-								}
+								onTriggerFailure(checkpoint, throwable);
 							}
+						} else {
+							if (checkpoint.isDiscarded()) {
+								onTriggerFailure(
+									checkpoint,
+									new CheckpointException(
+										CheckpointFailureReason.TRIGGER_CHECKPOINT_FAILURE,
+										checkpoint.getFailureCause()));
+							} else {
+								// no exception, no discarding, everything is OK
+								final long checkpointId = checkpoint.getCheckpointId();
+								snapshotTaskState(
+									timestamp,
+									checkpointId,
+									checkpoint.getCheckpointStorageLocation(),
+									request.props,
+									executions,
+									request.advanceToEndOfTime);
+
+								coordinatorsToCheckpoint.forEach((ctx) -> ctx.afterSourceBarrierInjection(checkpointId));
+
+								onTriggerSuccess();
+							}
+						}
 
-							return null;
-						},
-						timer));
+						return null;
+					},
+					timer)
+				.whenComplete((unused, error) -> {
+					if (error != null) {
+						if (!isShutdown()) {
+							failureManager.handleJobLevelCheckpointException(new CheckpointException(EXCEPTION, error), Optional.empty());
```

Comment:
It's a serious issue if it gets here. `CheckpointCoordinator` might be never recover even we fail the job. The state of `isTrigger` might not be reset, and it couldn't be reset forever. So we should reset the `isTriggering` to false first and then call the `failureManager`.

Reply:
I think if a failure occurs here, then there must a programming error or it is an artifact of a non properly shut down task. In the first case, I think we should fail hard. In the latter case (e.g. `if (isShutdown())` we should ignore the exception/log it on debug).

Reply:
Consequently, I'm not convinced that we need to call the `failureManager` here.

Reply:
Which exceptions do you expect here to appear?

Reply:
> Which exceptions do you expect here to appear?

Programming errors here like we had recently.

> I think we should fail hard. 

Do you mean `System.exit` by "fail hard"? I think it is not suitable because it:
- hides error details
- hides error itself in failover setup?
- prevents resource cleanup
- fails other jobs

Just throwing an exception here also won't work because this will only fail the timer thread.

Reply:
If we are only talking about programming errors, then I believe we should call `System.exit` (== fail hard) because programming errors usually leave the system in a corrupted state. 

Why do you think that it hides error details?

Not doing resource clean up in a failure case is acceptable.

Failing other jobs if the process has been corrupted is fine as well.

Reply:
> Why do you think that it hides error details?

Because `System.exit` can leave log buffers containing error details unflushed.

Plus 
> hides error itself in failover setup?

I.e. if programming error corrupts state then job restart can be undesirable.
Just crashing doesn't prevent restart, but failing the job does.

Reply:
I think doing `System.exit` is not nice practice in general. Flushing logs is one thing, but there might be some other exception handling code that wouldn't be execute as expected. Maybe some production code, like cleaning up temp files, maybe some debug code someone was counting on. Also it's confusing, as it's not the normal/usual way how Flink fails (with `System.exit` I guess there would be no message in the error log, but only on stderr?).

In many other places we have `checkState`s for programming errors.

If failing single job is not enough, we should fail whole JM, but via some exception. However I'm not buying the argument why should we fail other jobs as `CheckpointCoordinator` is bound to a single job. For example we are not failing whole TM if there was a `checkState` or `NPE` failure in the runtime code of a single job.

Reply:
>  Also it's confusing, as it's not the normal/usual way how Flink fails (with System.exit I guess there would be no message in the error log, but only on stderr?).

I fear that your statement is not true @pnowojski. We actually try to set wherever possible `FatalExitExceptionHandler` as the uncaught exception handler for every thread we start (it is not perfect yet). Hence, if there is an exception which bubbles up in the runtime code (e.g. `checkState`), then it will shut down the JVM. I believe that this is the right thing to do as these situations constitute a programming error and it is impossible to recover from all kinds of programming errors. Moreover, in case of programming errors or other fatal errors it is in general not possible to properly clean up. Just imagine an OOM when triggering the clean up code, for example.

Moreover, we are talking about code where the assumption is that all recoverable exceptions are properly being handled before it is called. Hence, it only makes sense to fail hard if this assumption is wrong.

> I.e. if programming error corrupts state then job restart can be undesirable.
Just crashing doesn't prevent restart, but failing the job does.

This is true, but at the moment we don't really support that you can directly fail a job w/o restarts, hence, the next best thing is to fail the process so that people see the problem.

> Because System.exit can leave log buffers containing error details unflushed.

Yes this can potentially happen. So far, however, this has never been a problem.

Reply:
If you don't really know what kind of exceptions can occur after `handleAsync` is called, then this is a different story. In this case I would advise to only log the error. However, then I would strongly suggest to get the `CheckpointCoordinator` in order soon because it is not healthy that random exceptions are allowed to fly around.

Reply:
I've reverted to `System.exit` after an offline discussion with @pnowojski (the motivation is to have more a simpler behavior and changeset given the time pressure of 1.11)